### PR TITLE
Correct argument list for DownInterpolate4HistoryValues function calls

### LINKED ISSUE #9566: Wrong argument order on some DownInterpolate4HistoryValues function calls
Issue overview
--------------
The function DownInterpolate4HistoryValues is used to update/revert thermal history values when the simulation reduces the system time step. This will revert thermal histories to what they were before the system time step was reduced so that new calculations can be performed as if a fresh new time step were simulated. The function arguments should be in a specific order. The first 2 arguments are the old and new time step values. The following e arguments are the noral history terms. The next argument is the value of a variable as the end of the current time step (time step not yet reduced). The final DS* arguments are the storage locations for the reduced time step (Down Shift) variables that will be used while the system time step is at a reduced time interval.

For example this call has the correct argument order:

                DownInterpolate4HistoryValues(PriorTimeStep,
                                              TimeStepSys,
                                              state.dataHeatBalFanSys->XMAT(ZoneNum),
                                              state.dataHeatBalFanSys->XM2T(ZoneNum),
                                              state.dataHeatBalFanSys->XM3T(ZoneNum),
                                              state.dataHeatBalFanSys->XM4T(ZoneNum),
                                              state.dataHeatBalFanSys->XM4T(ZoneNum), (not used)
                                              MAT(ZoneNum),
                                              state.dataHeatBalFanSys->DSXMAT(ZoneNum),
                                              state.dataHeatBalFanSys->DSXM2T(ZoneNum),
                                              state.dataHeatBalFanSys->DSXM3T(ZoneNum),
                                              state.dataHeatBalFanSys->DSXM4T(ZoneNum));

Several other calls have incorrect argument order used in the call:

                DownInterpolate4HistoryValues(PriorTimeStep,
                                              TimeStepSys,
                                              MAT(ZoneNum),
                                              state.dataHeatBalFanSys->XMAT(ZoneNum),
                                              state.dataHeatBalFanSys->XM2T(ZoneNum),
                                              state.dataHeatBalFanSys->XM3T(ZoneNum),
                                              state.dataHeatBalFanSys->XM4T(ZoneNum),
                                              MAT(ZoneNum),
                                              state.dataHeatBalFanSys->DSXMAT(ZoneNum),
                                              state.dataHeatBalFanSys->DSXM2T(ZoneNum),
                                              state.dataHeatBalFanSys->DSXM3T(ZoneNum),
                                              state.dataHeatBalFanSys->DSXM4T(ZoneNum));

Incorrect arguement order will not allow the program to revert the state of the program to where it was before the system time step was reduced and result in inaccurate results (small differences).

Work around: not really a work around but if Timestep and minimum system time step are set equal, these calls should have no impact on the result.

    Timestep,4;

    and

    ConvergenceLimits,
      0, \field Minimum System Timestep (uses Timestep as minimum value)

    or

    ConvergenceLimits,
      15, \field Minimum System Timestep (i.e., 4 15-minute intervals per hour)

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by rraustad**
Pull request overview
---------------------
 - Fixes #9566
 - Coordination of thermal history terms is not uniform. This will result in slight differences in results when the simulation chooses to reduce the system time step.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: tst/EnergyPlus/unit/ZoneTempPredictorCorrector.unit.cc
Comment (rraustad) [inline tst/EnergyPlus/unit/ZoneTempPredictorCorrector.unit.cc:1798]: This is an example of how the DownInterpolate4HistoryValues function should work. During the simulation at a 0.25 hour simulation, the history terms are 1, 2, 3, 4. When the simulation comes to a time step where the zone temp changes more than the limit, the simulation will down-shift. When that happens the history terms are reverted AND converted to the new time step, in this case 0.125 hours. It does not matter what the value of the variable is (in this case 5) at the end of the time step, when the simulation determines the simulation will down-shift. All values are reverted to what they were at the end of the last time step and history terms are interpolated to provide a history series in the same time domain as the shortened time step simulation.

File: src/EnergyPlus/ZoneContaminantPredictorCorrector.cc
Comment (rraustad) [inline src/EnergyPlus/ZoneContaminantPredictorCorrector.cc]: As an example of what this change represents. In this case, if the simulation down-shifted, the value of ZoneAirCO2 would be set equal to ZoneAirCO2 and would not get reverted to what it was at the last time step. In addition to that, the new history terms would be referenced from the value of ZoneAirCO2 at the end of the time step that would be resimulated at shorter time steps (i.e., the longer simulation results should be deleted/reverted and not used). This should not cause noticable differences in results but there would certainly be numeric differences from the correct simulation results.

**Comment by rraustad** [comment]
I am surprised there is only 1 diff from RoomAirflowNetwork. I thought there would be diffs in CO2 and contaminant example files. And for sure other normal example files since the calls for zone T and w also had mistakes, albeit in the CorrectZoneAirTemp function at the end of the time step.

**Comment by mjwitte** [comment]
> I am surprised there is only 1 diff from RoomAirflowNetwork. I thought there would be diffs in CO2 and contaminant example files. And for sure other normal example files since the calls for zone T and w also had mistakes, albeit in the CorrectZoneAirTemp function at the end of the time step.

I agree. Looking at HeatPumpProportionalControl_DCV locally, there are several times where the system downsteps, but no diffs in final results. Any thoughts why? Maybe CO2 doesn't use all the terms?

**Comment by rraustad** [comment]
@mjwitte I have not had time to look. I would rather focus first on typical example files that have a down-shift and no diffs to see if they hit one of the changed calls.

**Comment by mjwitte** [comment]
In the case of contaminants, it does not hit the modified call to `DownInterpolate4HistoryValues` because `ShortenTimeStepSys` is always set to `false` before the correct step [here](https://github.com/NREL/EnergyPlus/blob/262f9e5184a90cd6439f25599f3b71c9ea91fdae/src/EnergyPlus/HVACManager.cc#L354-L355). Going to check the blame on that. If that's the case, then why do the correct functions even have this call?

**Comment by mjwitte** [comment]
For the record, here are the [comit](https://github.com/NREL/EnergyPlusArchive/commit/b225c4e4f3a4d49f06156aedf1b3e3a598139213) and issue [838](https://github.com/NREL/EnergyPlus/issues/838) where this was changed. So, it's quite possible that the entire `if (ShortenTimeStepSys) {` in the correct functions can be removed. Next stop is to look at the coverage to see if these are ever hit.

**Comment by rraustad** [comment]
I was thinking the same thing. The reason to down-shift is to simulate beginning with predict. Why down-shift again at the end of a time step, shortened or otherwise.

**Comment by mjwitte** [comment]
Sure enough those blocks are red in the integration coverage details. Rip 'em out. And maybe add a comment line and then that parameter can be removed from the correct functions.

File: src/EnergyPlus/ZoneContaminantPredictorCorrector.cc
Comment (mjwitte) [inline src/EnergyPlus/ZoneContaminantPredictorCorrector.cc]: Oops - `PriorTimeStep` shouldn't be here. I'll fix that shortly.

File: src/EnergyPlus/ZoneContaminantPredictorCorrector.cc
Comment (mjwitte) [inline src/EnergyPlus/ZoneContaminantPredictorCorrector.cc:2378]: Byebye!

**Review by mjwitte — APPROVED** [review]
This all looks good. Deleted code is a good thing, especially dead code.

Will wait on CI to come back before merging.

File: src/EnergyPlus/ZoneContaminantPredictorCorrector.cc
Comment (rraustad) [reply src/EnergyPlus/ZoneContaminantPredictorCorrector.cc]: 1) PriorTimeStep should not be there either, 2) I hope you caught this, and 3) why did this compile?

**Comment by rraustad** [comment]
This does look correct to me. @mjwitte fixed a mistake and overall things line up with what is expected. If the simulation determines that a reduced time step is needed then revert history terms and resim. That should happen in the predict step, right off the bat. Never interpolate at the correct step, which was proven to never be used. The unit test exercises this concept. The only thing to be sure of is that all models that use this concept 1) use the same methodology, 2) use the same code (no dups, ZoneContaminantPredictorCorrector), and 3) actually use the correct accounting (RoomAirModelAirflowNetwork should call DownInterpolate4HistoryValues?). This is at least a good first step.

File: src/EnergyPlus/ZoneContaminantPredictorCorrector.cc
Comment (Myoldmopar) [inline src/EnergyPlus/ZoneContaminantPredictorCorrector.cc:137]: Fewer arguments :heart:

File: src/EnergyPlus/ZoneContaminantPredictorCorrector.cc
Comment (Myoldmopar) [reply src/EnergyPlus/ZoneContaminantPredictorCorrector.cc:2378]: Less code! :heart_eyes:

File: tst/EnergyPlus/unit/ZoneTempPredictorCorrector.unit.cc
Comment (Myoldmopar) [reply tst/EnergyPlus/unit/ZoneTempPredictorCorrector.unit.cc:1798]: Love it.

**Review by Myoldmopar — APPROVED** [review]
Great!  And about a hundred more lines gone!

**Comment by Myoldmopar** [comment]
This seems fully ready to go.  CI is already pending on it, so I'll give it another 30 minutes or so to wrap up a few builds, but I suspect it will merge shortly after that.

**Comment by Myoldmopar** [comment]
Happy happy, let's merge this one.  Thanks @rraustad and review crew!

**Pull request merged**
